src/main/scala/au/com/agiledigital/toolform/tasks/generate/docker/GenerateDockerComposeV3.scala

```diff
@@ -0,0 +1,175 @@
+       .exec(context),
+     ())
+  }
+
+  private def writeComponents(projectId: String, components: List[Component]) = State[WriterContext, Unit] { context =>
+    (components
+       .traverseU(component => writeComponent(projectId, component))
+       .exec(context),
+     ())
+  }
+
+  private def writeResources(resources: List[Resource]) = State[WriterContext, Unit] { context =>
+    (resources
+       .traverseU(resource => writeResource(resource))
+       .exec(context),
+     ())
+  }
+
+  private def writeSubEdge(projectId: String, subEdgeDef: SubEdgeDef) = {
+    val serviceName = subEdgeServiceName(projectId, subEdgeDef)
+    val imageName   = subEdgeImageName(projectId, subEdgeDef)
+    for {
+      _ <- write(s"$serviceName:")
+      i <- indent()
+      _ <- for {
+        _ <- write(s"image: $imageName")
+        _ <- write(s"restart: always")
+        _ <- write(s"ports:")
+        _ <- write(subEdgePortDefinition(subEdgeDef))
+      } yield ()
+      _ <- resetIndent(i)
+    } yield ()
+  }
+
+  private def writeComponent(projectId: String, component: Component) = {
+    val serviceName = componentServiceName(component)
+    val imageName   = componentImageName(projectId, component)
+    for {
+      _ <- write(s"$serviceName:")
+      i <- indent()
+      _ <- for {
+        _ <- write(s"image: $imageName")
+        _ <- write(s"restart: always")
+        _ <- writeComponentLabels(component)
+        _ <- writeEnvironmentVariables(component)
+        _ <- writePorts(component)
+      } yield ()
+      _ <- resetIndent(i)
+    } yield ()
+  }
+
+  private def writeComponentLabels(component: Component) =
+    for {
+      _ <- write("labels:")
+      i <- indent()
+      _ <- for {
+
+        _ <- write(s"source.path: \042${component.path}\042")
+        _ <- write("project.artefact: \"true\"")
+      } yield ()
+      _ <- resetIndent(i)
+    } yield ()
+
+  private def writeResource(resource: Resource) = {
+    val serviceName = resourceServiceName(resource)
+    val imageName   = resourceImageName(resource)
+    for {
+      _ <- write(s"$serviceName:")
+      i <- indent()
+      _ <- for {
+
+        _ <- write(s"image: $imageName")
+        _ <- write(s"restart: always")
+        _ <- writeEnvironmentVariables(resource)
+        _ <- writePorts(resource)
+      } yield ()
+      _ <- resetIndent(i)
+    } yield ()
+  }
+
+  private def writeEnvironmentVariables(service: Service) = service.environment match {
+    case Some(environmentEntries) =>
+      for {
+        _ <- write("environment:")
+        _ <- State[WriterContext, Unit] { context =>
+          {
+            (environmentEntries.toList
+               .map((entry) => formatEnvironment(entry))
+               .traverseU(write)
+               .exec(context),
+             ())
+          }
+        }
+      } yield ()
+    case None => identity
+  }
+
+  private def writePorts(service: Service) =
+    service.exposedPorts match {
```

Comment:
Ditto.

Reply:
Fixed by specifying a default value for environment and exposed ports in the case class.

This has now be simplified to an `if` statement checking if the list/map is empty or not.